File: src/main/java/com/babylonml/backend/training/operations/HadamardProduct.java
package com.babylonml.backend.training.operations;

import com.babylonml.backend.cpu.TensorOperations;
import com.babylonml.backend.training.execution.TensorPointer;
import com.babylonml.backend.cpu.VectorOperations;


import org.jspecify.annotations.NonNull;
import org.jspecify.annotations.Nullable;

import java.util.Objects;

public final class HadamardProduct extends AbstractOperation {
    @Nullable
    private TensorPointer leftOperandPointer;

    @Nullable
    private TensorPointer rightOperandPointer;

    private final boolean requiresDerivativeChainValue;

    private final int[] maxShape;

    public HadamardProduct(Operation leftOperation, Operation rightOperation) {
        super(leftOperation, rightOperation);

        this.maxShape = TensorOperations.calculateMaxShape(leftOperation.getMaxResultShape(),
                rightOperation.getMaxResultShape());
        this.requiresDerivativeChainValue = leftOperation.requiresBackwardDerivativeChainValue() ||
                rightOperation.requiresBackwardDerivativeChainValue();
    }

    @Override
    public int @NonNull [] getMaxResultShape() {
        return maxShape;
    }

    @Override
    public @NonNull TensorPointer forwardPassCalculation() {
        Objects.requireNonNull(leftOperation);
        Objects.requireNonNull(rightOperation);

        leftOperandPointer = leftOperation.forwardPassCalculation();
        rightOperandPointer = rightOperation.forwardPassCalculation();

        return broadcastIfNeeded(leftOperandPointer, rightOperandPointer, forwardMemoryAllocator,
                (firstTensor, secondTensor, result) ->
                        VectorOperations.vectorToVectorElementWiseMultiplication(firstTensor.buffer(), firstTensor.offset(),
                                secondTensor.buffer(), secondTensor.offset(), result.buffer(), result.offset(),
                                TensorOperations.stride(result.shape())));
    }

    @Override
    public @NonNull TensorPointer leftBackwardDerivativeChainValue() {
        Objects.requireNonNull(derivativeChainPointer);
        Objects.requireNonNull(rightOperandPointer);

        return reduceIfNeeded(derivativeChainPointer, rightOperandPointer, backwardMemoryAllocator,
                (derivativeTensor, rightTensor, resultTensor) ->
                        VectorOperations.vectorToVectorElementWiseMultiplication(

                                derivativeTensor.buffer(), derivativeTensor.offset(),
                                rightTensor.buffer(), rightTensor.offset(),
                                resultTensor.buffer(), resultTensor.offset(),
                                TensorOperations.stride(resultTensor.shape())));
    }

    @Override
    public @NonNull TensorPointer rightBackwardDerivativeChainValue() {
        Objects.requireNonNull(derivativeChainPointer);
        Objects.requireNonNull(leftOperandPointer);

        return reduceIfNeeded(derivativeChainPointer, leftOperandPointer, backwardMemoryAllocator,
                (derivativeTensor, leftTensor, resultTensor) ->
                        VectorOperations.vectorToVectorElementWiseMultiplication(
                                derivativeTensor.buffer(), derivativeTensor.offset(),
                                leftTensor.buffer(), leftTensor.offset(),
                                resultTensor.buffer(), resultTensor.offset(),
                                TensorOperations.stride(resultTensor.shape()))
        );
    }

    @Override
    public int @NonNull [][] getForwardMemoryAllocations() {
        return new int[][]{
                maxShape
        };
    }

    @Override
    public int @NonNull [][] getBackwardMemoryAllocations() {
        return new int[][]{
                maxShape,
                maxShape
        };
    }

    @Override
    public boolean requiresBackwardDerivativeChainValue() {
        return requiresDerivativeChainValue;
    }
}

File: src/main/kotlin/com/babylonml/backend/training/operations/HadamardProduct.kt
package com.babylonml.backend.training.operations

import com.babylonml.backend.common.CommonTensorOperations
import com.babylonml.backend.cpu.VectorOperations
import com.babylonml.backend.training.execution.TensorPointer
import it.unimi.dsi.fastutil.ints.IntImmutableList

class HadamardProduct(leftOperation: Operation, rightOperation: Operation) :
    AbstractOperation(leftOperation, rightOperation) {
    private var leftOperandPointer: TensorPointer? = null
    private var rightOperandPointer: TensorPointer? = null

    private val maxShape: IntImmutableList =
        CommonTensorOperations.calculateMaxShape(
            leftOperation.maxResultShape,
            rightOperation.maxResultShape
        )

    override val maxResultShape: IntImmutableList
        get() = maxShape

    override fun forwardPassCalculation(): TensorPointer {
        leftOperandPointer = leftPreviousOperation!!.forwardPassCalculation()
        rightOperandPointer = rightPreviousOperation!!.forwardPassCalculation()

        return broadcastIfNeeded(
            leftOperandPointer!!, rightOperandPointer!!, forwardMemoryAllocator
        ) { firstTensor: TensorPointer, secondTensor: TensorPointer, result: TensorPointer ->
            VectorOperations.vectorToVectorElementWiseMultiplication(
                firstTensor.buffer(), firstTensor.offset(),
                secondTensor.buffer(), secondTensor.offset(), result.buffer(), result.offset(),
                CommonTensorOperations.stride(result.shape)
            )
        }
    }

    override fun leftBackwardDerivativeChainValue(): TensorPointer {
        return reduceIfNeeded(
            derivativeChainPointer!!, rightOperandPointer!!, backwardMemoryAllocator
        ) { derivativeTensor: TensorPointer, rightTensor: TensorPointer, resultTensor: TensorPointer ->
            VectorOperations.vectorToVectorElementWiseMultiplication(
                derivativeTensor.buffer(), derivativeTensor.offset(),
                rightTensor.buffer(), rightTensor.offset(),
                resultTensor.buffer(), resultTensor.offset(),
                CommonTensorOperations.stride(resultTensor.shape)
            )
        }
    }

    override fun rightBackwardDerivativeChainValue(): TensorPointer {
        return reduceIfNeeded(
            derivativeChainPointer!!, leftOperandPointer!!, backwardMemoryAllocator
        ) { derivativeTensor: TensorPointer, leftTensor: TensorPointer, resultTensor: TensorPointer ->
            VectorOperations.vectorToVectorElementWiseMultiplication(
                derivativeTensor.buffer(), derivativeTensor.offset(),
                leftTensor.buffer(), leftTensor.offset(),
                resultTensor.buffer(), resultTensor.offset(),
                CommonTensorOperations.stride(resultTensor.shape)
            )
        }
    }

    override val forwardMemoryAllocations: List<IntImmutableList>
        get() = listOf(maxShape)

    override val backwardMemoryAllocations: List<IntImmutableList>
        get() = listOf(maxShape, maxShape)

    override val requiresBackwardDerivativeChainValue: Boolean by lazy {
        leftPreviousOperation!!.requiresBackwardDerivativeChainValue ||
                rightPreviousOperation!!.requiresBackwardDerivativeChainValue
    }
}
